### ISSUE TITLE
Plant_trial_scoring_template-feedback

### ISSUE THREAD

**Issue by teatree1212** [TGAC/brassica#608]
A after feedback:  a "comments" column should be added to the plant_trial_scoring_template. It should allow for free text and would always appear as the final header in the template.

**Comment by Nuanda** [comment]
This "comments" value will be associated with the Plant Scoring Unit represented by the given CSV row. Is this what you mean? In another place somewhere in GitHub you have mentioned "comments" for Trait Scores...

**Comment by teatree1212** [comment]
I remember mentioning it somewhere but couldn't find it. This is actually a tricky one.. I know people want to comment on certain observations they have while scoring the traits, whether they anticipate that this trait score may not be quite correct for this and that reason would require recording additional information as  a comment (PSU)- see picture.
But sometimes they just want to record additional observations on the Plant itself (Plant Scoring Unit)
<img width="979" alt="screen shot 2016-07-26 at 10 56 37" src="https://cloud.githubusercontent.com/assets/17452217/17133855/c2b5bd38-531f-11e6-998f-2a96c0f1115d.png">


I see that with it being recorded in the TS.comments table, it would refer to a single trait, which however is appropriate in the case of what is shown in the picture.

Do you think it would be too complicated for the people to follow this logic:

When simply adding a comment to the "comments" column, this would have to be a general comment that applies to the PSU ( the plant itself) in general. it then has to be parsed to all TS.comments entries ( or the PSU.comments field).
One would write into the comments column: "Plant particularly stunted"

When wanting to add a comment on a specific trait, they would have to go about it in a certain fashion: mention the number of the cell and then the problem that arose when scoring this particular trait
For example in the screenshot, the first comment applied to Trait2 "Slightly damaged": that would have to be applied to this specific trait and parsed to the TS.comments table.
So one would have to write "7: slightly damaged". with multiple traits affected: "6: Double stem, 8: Double stem, 10: Double stem")


There may be better ways of doing it, it is just a thought, lets discuss it here.

**Comment by Nuanda** [comment]
One option is to let the user extend the submitted CSV in two ways:
 - add another, optional `Comments` column right after the trait scores column which the comment in question is related to (they will be parsed and saved in TraitScore.comments column)
 - add yet another `Scoring unit comments` column as the last one in the CSV file, where PSU-related comments would be placed (they will be parsed and stored in PlantScoringUnit.comments column).

However, in your example case this would produce 4 additional `Comments` columns (one after each of trait scoring columns 1,2,3 and 5) only to put 5 comments in them. So it results in a pretty sparse CSV file.

Is the screenshot above the "regular" way how they (i.e. the plant trial people) solve the comments issue? If so, perhaps we should somehow accommodate it...

**Comment by teatree1212** [comment]
I haven't seen this before in other sheets. But I think their solution is quite good, once people understand how it is done.

**Comment by Nuanda** [comment]
The problem with this solution is a "sociotechnical" one. **If** the content of the final Commens column be parsed and interpreted so *pieces* of the comment are saved with particular *scores*, not just the PSU, then it introduces a kind of grammar for users. This has the following drawbacks:
 - it will add a considerable new section to the scoring template "manual" that people need to read and understand (and we know it is already long) - but that's not a real problem
 - it will cause a lot of errors with people making all kinds of "syntax errors" with this solution -> some of these may be alleviated with complex parsing solution, other we will never think about beforehand :).

From your experience with current plant trials - are such comments numerous? If there are like 5 comments for the entire trial, maybe we should add a specialized widget/form in Step 5 for this? It would ask the user to pick a specific PSU, then a specific Trait (and, perhaps, a tech repl number), and it will allow him/her to type in a comment for this particular trait score.

Of course, if for a single trial there are many more comments, this will not result in a good user experience.

**Comment by teatree1212** [comment]
I like the widget idea a lot. I think it might be the easiest way. 

However:
this is a -maybe rare but we see possible- case, where there are quite many comments made 
[widget_idea_comments.xlsx](https://github.com/TGAC/brassica/files/449605/widget_idea_comments.xlsx)

The amount of comments is quite high there.
The problem is that they also have a " key" for their comments GB= Green Bud. So that must be noted somewhere.  This makes this "comments" column almost a trait column


And in this example instance I thought it would be best  to make a "trait" such as the image trait, called "plot observations of note" -In this case:
**Plot observations of note**
During trait scoring of other kind, anything of note about the plot was scored, eg. GB (green bud), EL (elongating), YB (yellow bud).

What I am worried about is that people confuse this trait with a comment  column. And that they also use it for something other than commenting on GB EL YB, which has a key that is however not described in that particular trait. This would mean that people would have to create a "plot observations of note_ 2" trait and define the observations differently. 

...
Which is why I like the Crop Ontology model so much, where you can have the same trait, but defined by different methods. But it will be some time until we will integrate with them.


What are your thoughts on the above?

**Comment by teatree1212** [comment]
Would it be possible to 
- display the .csv-uploaded data in an interactive window during the upload process 
-  the person would be able to click on a field in the spreadsheet and add comments to it that way ( similar to in an excel spreadsheet)? 
-> the comments will be parsed to the respective fields in the database, i.e. if there is a comment on a particular measurement value, this certain field in the spreadsheet, by clicking on it, it would change colour and a window appears, for the user to add a comment. by clicking  on another single field, the user should be able to select the previous comment made so that filling it out doesn't require too much time. ( or ( the user can mark  multiple measurement fields in the spreadsheet for which a comment is valid -> this would be parsed to Trait_scores.comments). Comments would be related to the measurement itself, that could say something like "sampling machine was turned off for a few seconds, measurement may not be correct, as not calibrated"

->The user could also click on a certain Plant scoring unit name -> would be parsed to PSU.comments, and add comments like " this replicate looked very sad in comparison, measurements may be unreflective of the line's characteristics" or "sparse and small"


-> other headers could not be manipulated, but we even could let the user add comments to the individual ( or all) plant_accessions, if we wanted

As this may still be quite complicated:
I think here it would be best to maybe have a short video that the user can click on immediately when looking at this spreadsheet, which explains what can be done (If this was the way we implemented it, I can make the video and you can add a link to it to the section)
you could even have a little text box saying: "you are commenting on ALL replicates" or "you are commenting on replicate x "or "you are commenting on trait scoring value y for the trait z"
-> so that the user definitely understands what the comment is directed to.

But what do you think? I think it would be easiest for the user, as this is very similar to commenting on an excel spreadsheet.

**Comment by Nuanda** [comment]
I'll be honest this might be quite a bit of work. I.e. it will cost a lot of developer time. Sure, it looks natural for users (though remember pagination and vertical scrolling and things like that... they will add to the UI complexity). Drawback is that people would need to retype, or copy one-by-one all comments (which they already have in their excel files) - they might be nonplussed about that. They might think "I have already did that in my Excel spreadsheet, now I need to do that with this 'web spreadsheet' again...".

While when they are forced to split their comments into score-related ones and psu-related ones, they might simple copy-over entire excel columns and quickly remove irrelevant parts.

But aren't these GB, YB comments more about PSUs (plant samples) than about a particular score?

**Comment by teatree1212** [comment]
okay, I understand. Pleas make sure this idea is then removed from the extension quote if @wjurkowski hasn't already done so.

Yes, in the case above GV YB are related to the PSU.  The document above is just an example that sometimes there are a lot of comments present. 

However, in the above file there even are two different types of comments  , where one is a string, but one is actually part of a scoring scale, with certain keys (e.g. GB= Green Bud) on a scale from GB< GB-EL < YB < (more are possible) < to Flowering ( FL) or beyond. This is why I thought of the *plot observations of note* trait variable similar to the image in the comment further above.
_--> this is where a trait definition, data model and data submission and parsing problem come together in one column.._
What do you think about the plot observations of note trait? 

In case we can actually separate such scorings from proper "comments" ( and people need to understand that somehow - sociotechnical problem here, top)

we can go back to the widget idea, which I want to comment on in the context of your most recent comment 
-> would the user not - here as well-think  :"I  already did that in my Excel spreadsheet, now I need to do that with this [widget] again...".

**Comment by wjurkowski** [comment]
Could we first discuss rationale behind the need to add comments to spreadsheet?

**Comment by Nuanda** [comment]
@teatree1212 You are right wrt the widget solution having the same "once again" problem. I think I still like the additional CSV column(s) idea the best.

**Comment by Nuanda** [comment]
After a discussion with @wjurkowski today we have decided to go with the solution based on additional, optional columns in the uploaded CSV document.

**Comment by kammerer** [comment]
See also #732
